cogs/study_groups.py

```diff
@@ -0,0 +1,135 @@
+import asyncio
+import re
+from typing import Union
+
+from discord import Guild, Role, Member, User
+from discord.ext.commands import Cog, Bot, command, Context, group, bot_has_guild_permissions, BadArgument
+from discord_components import DiscordComponents, Interaction, Select, \
+    SelectOption
+
+from cogs.botStatus import listener
+from cogs.util.study_subject_util import StudySubjectUtil
+from core.globalEnum import SubjectsOrGroupsEnum, CollectionEnum, ConfigurationNameEnum
+from mongo.primitiveMongoData import PrimitiveMongoData
+from mongo.subjectsorgroups import SubjectsOrGroups
+
+
+class StudyGroups(Cog):
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.match = re.compile(r"([a-z]+)([0-9]+)", re.I)
+        self.db = SubjectsOrGroups(self.bot, SubjectsOrGroupsEnum.GROUP)
+        self.startup = True
+        self.groups = {}
+
+    @listener()
+    async def on_ready(self):
+        if self.startup:
+            DiscordComponents(self.bot)
+            self.startup = False
+
+    def cog_unload(self):
+        pass
+
+    @command()
+    async def study(self, ctx: Context):
+        guild: Guild = ctx.guild
+        member: Union[Member, User] = ctx.author
+        groups = sorted({guild.get_role(document.role_id) for document in await self.db.find({})},
+                        key=lambda role: role.name)
+        groups: list[Role]
+        # limit self assigment to semester one and two
+        groups = [_group for _group in groups if (int((re.match(self.match, _group.name).groups())[1]) < 3)]
+
+        group_names = {str((re.match(self.match, _role.name).groups())[0])
+                       for _role in groups}
+        group_names_options = Select(
+            placeholder="Select your group",
+            options=[SelectOption(label=name, value=name) for name in group_names])
+
+        group_semester = {str((re.match(self.match, _role.name).groups())[1]) for _role in groups}
+        group_semester_options = Select(
+            placeholder="Select your semester",
+            options=[SelectOption(label=semester, value=semester) for semester in group_semester])
+
+        await ctx.reply(content="Please select **one** of the following groups.",
+                        components=[group_names_options, group_semester_options])
+
+        role: Role = await self.get_role(groups, group_names, group_semester)
+        await member.add_roles(role)
+
+    async def get_role(self, groups: list[Role], group_names, group_semester) -> Role:
+        a, b = await asyncio.gather(
```

Comment:
Since we don't know what we'll receive first, we can't accurately identify the content.